br/pkg/membuf/buffer.go

```diff
@@ -130,64 +125,130 @@ func (p *Pool) TotalSize() int64 {
 
 // Buffer represents the reuse buffer.
 type Buffer struct {
-	pool      *Pool
-	bufs      [][]byte
-	curBuf    []byte
-	curIdx    int
-	curBufIdx int
-	curBufLen int
+	pool          *Pool
+	blocks        [][]byte
+	blockCntLimit int
+	curBlock      []byte
+	curBlockIdx   int
+	curIdx        int
 }
 
-// addBuf adds buffer to Buffer.
-func (b *Buffer) addBuf() {
-	if b.curBufIdx < len(b.bufs)-1 {
-		b.curBufIdx++
-		b.curBuf = b.bufs[b.curBufIdx]
-	} else {
-		buf := b.pool.acquire()
-		b.bufs = append(b.bufs, buf)
-		b.curBuf = buf
-		b.curBufIdx = len(b.bufs) - 1
+// BufferOption configures a buffer.
+type BufferOption func(*Buffer)
+
+// WithMemoryLimit approximately limits the maximum memory size of this Buffer.
+// Due to it use blocks to allocate memory, the actual memory size is
+// blockSize*ceil(limit/blockSize).
+// In order to keep compatibility, it will only restrict AllocBytesWithSliceLocation.
+func WithMemoryLimit(limit uint64) BufferOption {
+	return func(b *Buffer) {
+		blockCntLimit := int(limit+uint64(b.pool.blockSize)-1) / b.pool.blockSize
+		b.blockCntLimit = blockCntLimit
+		b.blocks = make([][]byte, 0, blockCntLimit)
 	}
+}
 
-	b.curBufLen = len(b.curBuf)
-	b.curIdx = 0
+// NewBuffer creates a new buffer in current pool.
+func (p *Pool) NewBuffer(opts ...BufferOption) *Buffer {
+	b := &Buffer{
+		pool:          p,
+		blocks:        make([][]byte, 0, 128),
```

Comment:
Tiny improvement: maybe move this initialization after all `BufferOption` are applied? Thus we don't waste this slice if `WithMemoryLimit` exists.